src/client/debugger/extension/adapter/factory.ts

```diff
@@ -28,20 +23,29 @@ export class DebugAdapterDescriptorFactory implements IDebugAdapterDescriptorFac
     constructor(
         @inject(IInterpreterService) private readonly interpreterService: IInterpreterService,
         @inject(IApplicationShell) private readonly appShell: IApplicationShell,
-        @inject(IExperimentsManager) private readonly experimentsManager: IExperimentsManager,
-        @inject(IPythonExecutionFactory) private readonly executionFactory: IPythonExecutionFactory,
-        @inject(IPersistentStateFactory) private readonly stateFactory: IPersistentStateFactory,
-        @inject(IExtensions) private readonly extensions: IExtensions
-    ) {}
+        @inject(IExperimentsManager) private readonly experimentsManager: IExperimentsManager
+    ) { }
     public async createDebugAdapterDescriptor(session: DebugSession, executable: DebugAdapterExecutable | undefined): Promise<DebugAdapterDescriptor> {
         const configuration = session.configuration as (LaunchRequestArguments | AttachRequestArguments);
-        const pythonPath = await this.getPythonPath(configuration, session.workspaceFolder);
 
-        if (await this.useNewPtvsd(pythonPath)) {
-            // If logToFile is set in the debug config then pass --log-dir <path-to-extension-dir> when launching the debug adapter.
-            const logArgs = configuration.logToFile ? ['--log-dir', EXTENSION_ROOT_DIR] : [];
-            const ptvsdPathToUse = await this.getPtvsdPath(pythonPath);
-            return new DebugAdapterExecutable(`${pythonPath}`, [path.join(ptvsdPathToUse, 'adapter'), ...logArgs]);
+        if (this.experimentsManager.inExperiment(DebugAdapterNewPtvsd.experiment)) {
+            if (configuration.request === 'attach') {
+                const port = configuration.port ? configuration.port : 0;
+                if (port === 0) {
```

Comment:
Is it possible to have a configuration with port === 0?

```suggestion
                if (port === undefined) {
```

and further down
```ts
return new DebugAdapterServer(configuration.port, configuration.host);
```

Reply:
Users can put `port=0` in the launch json. It is not a valid value when connecting as a client.